### ISSUE TITLE
2.23 

### ISSUE THREAD

**Issue by liusy58** [liusy58/LearningCS#2]

**Comment by liusy58** [comment]
C++如何launch一个线程呢？

**Comment by liusy58** [comment]
```C++
#include<thread>
class background_task
{
public:
    void do_something()const{}
    void do_something_else()const{}
    void operator()() const
    {
        do_something();
        do_something_else();
    }
};

int main(){
    std::thread my_thread(background_task());
    
    
}

```

这其中包括了一个C++ most vexing parse的问题。这个函数的参数怎么确定的呢？


`T1 name(T2());`
这是一个 declaration statement。
既可视作声明了一个类型为 T1 名为 name 的 object，并且用一个类型为 T2 的 object 作为其 initilizer（即标准中所谓「an object declaration with a function-style cast as the initializer」），也可视作声明了一个返回值类型为 T1 名为 name 的函数，此函数有一个参数，参数类型为「指向返回值类型为 T2，参数为空的函数的指针」。

**Comment by liusy58** [comment]
3. 一定要在thread析构之前调用`join`或者`detach`，否则thread会直接调用`terminate`来直接终止程序。

**Comment by liusy58** [comment]
4. 如果使用了`detach`不等待函数，一定要注意线程是否持有创建该线程的函数所有的局部变量。

```C++
struct func
{
    int &i;
    func(int &i_) : i(i_) {}
    void operator()()
    {
        for (unsigned j = 0; j < 1000000; ++j)
        {
            do_something(i);
        }
    }
};
void oops()
{
    int some_local_state = 0;
    func my_func(some_local_state);
    std::thread my_thread(my_func);
    my_thread.detach();
}
```

**Comment by liusy58** [comment]
> 4. 如果使用了`detach`不等待函数，一定要注意线程是否持有创建该线程的函数所有的局部变量。
> 
> ```c++
> struct func
> {
>     int &i;
>     func(int &i_) : i(i_) {}
>     void operator()()
>     {
>         for (unsigned j = 0; j < 1000000; ++j)
>         {
>             do_something(i);
>         }
>     }
> };
> void oops()
> {
>     int some_local_state = 0;
>     func my_func(some_local_state);
>     std::thread my_thread(my_func);
>     my_thread.detach();
> }
> ```

解决办法： 将数据拷贝到线程中去，而不是共享数据。

**Comment by liusy58** [comment]
5. 为什么不能呢个多次调用`join`?

因为调用`join`还会回收该线程相关的内存，`std::thread`这个对象将不再与刚刚结束的线程对应，因此不能多次调用。

**Comment by liusy58** [comment]
6. 需要注意在`try catch`语句中`join`的调用的时刻。

```C++
struct func;
void f()
{
    int some_local_state = 0;
    func my_func(some_local_state);
    std::thread t(my_func);
    try
    {
        do_something_in_current_thread();
    }
    catch (...)
    {
        throw;
    }
    t.join();
}
```

这个就是有问题的，一旦有异常抛出，那么函数结尾的`join`函数会被跳过，因此会出现thread没有`join`和`detach`从而造成`terminate`的情况。
解决方法：

```C++
struct func;
void f()
{
    int some_local_state = 0;
    func my_func(some_local_state);
    std::thread t(my_func);
    try
    {
        do_something_in_current_thread();
    }
    catch (...)
    {
        t.join();
        throw;
    }
    t.join();
}
```

或者使用RAII（Resource Acquisition Is Initialization）定律。

```C++
class thread_guard
{
    std::thread &t;

public:
    explicit thread_guard(std::thread &t_) : t(t_) {}
    ~thread_guard()
    {
        if (t.joinable())
        {
            t.join();
        }
    }
    thread_guard(thread_guard const &) = delete;
    thread_guard &operator=(thread_guard const &) = delete;
};
struct func;

void f()
{
    int some_local_state = 0;
    func my_func(some_local_state);
    std::thread t(my_func);
    thread_guard g(t);
    do_something_in_current_thread();
}
```

**Comment by liusy58** [comment]
> 

7. 在`thread_guard`中为什么将拷贝构造函数和赋值函数定义为`deleted`的？

**Comment by liusy58** [comment]
8. 如何给线程函数传递参数？

当创建一个线程，传递给`thread`的参数默认先被拷贝到可以被刚创建线程可以访问到的内部区域，然后被当作右值传递给可调用的对象或者函数。

这种情况需要注意几点问题：

a. 当心dangling pointer

```C++
void f(int i, std::string const &s);
void oops(int some_param)
{
    char buffer[1024];
    sprintf(buffer, "%i", some_param);
    std::thread t(f, 3, buffer);
    t.detach();
}
```

在这个例子中，buffer仍然是一个指针，就算被拷贝到可以访问的内部区域，但是当oop这个函数结束后，buffer所指向的区域是非法的，从而造成了错误。

改正方法：

```C++
void f(int i, std::string const &s);
void oops(int some_param)
{
    char buffer[1024];
    sprintf(buffer, "%i", some_param);
    std::thread t(f, 3, std::string(buffer));
    t.detach();
}
```

**Comment by liusy58** [comment]
9. 如何给线程构造函数传递成员函数来调用呢？

跟`bind`一样的[语义](https://www.jianshu.com/p/f191e88dcc80)。
```C++
class X
{
public:
    void do_lengthy_work();
};
X my_x;
std::thread t(&X::do_lengthy_work, &my_x);
```

**Comment by liusy58** [comment]
10. 如何给线程构造函数传递引用参数呢?
```C++
std::thread t(update_data_for_widget,w,std::ref(data));
```

**Comment by liusy58** [comment]
11. 如何得到一个thread的id呢？

```C++
std::thread::id master_thread;
void some_core_part_of_algorithm()
{
  if(std::this_thread::get_id()==master_thread)
  {
    do_master_thread_work();
  }
  do_common_work();
}
```
